util/resourcegrouptag/resource_group_tag.go

```diff
@@ -52,3 +60,149 @@ func DecodeResourceGroupTag(data []byte) (sqlDigest []byte, err error) {
 	}
 	return tag.SqlDigest, nil
 }
+
+// GetResourceGroupLabelByKey determines the label of the resource group based on the content of the key.
+func GetResourceGroupLabelByKey(key []byte) tipb.ResourceGroupTagLabel {
+	if len(key) == 0 {
+		return tipb.ResourceGroupTagLabel_ResourceGroupTagLabelUnknown
+	}
+	_, _, isRow, err := decodeKeyHead(key)
+	if err != nil {
+		return tipb.ResourceGroupTagLabel_ResourceGroupTagLabelUnknown
+	}
+	if isRow {
+		return tipb.ResourceGroupTagLabel_ResourceGroupTagLabelRow
+	}
+	return tipb.ResourceGroupTagLabel_ResourceGroupTagLabelIndex
+}
+
+// GetFirstKeyFromRequest gets the first Key of the request from tikvrpc.Request.
+func GetFirstKeyFromRequest(req *tikvrpc.Request) (firstKey []byte) {
+	if req == nil {
+		return
+	}
+	switch req.Req.(type) {
+	case *kvrpcpb.GetRequest:
+		r := req.Req.(*kvrpcpb.GetRequest)
+		if r != nil {
+			firstKey = r.Key
+		}
+	case *kvrpcpb.BatchGetRequest:
+		r := req.Req.(*kvrpcpb.BatchGetRequest)
+		if r != nil && len(r.Keys) > 0 {
+			firstKey = r.Keys[0]
+		}
+	case *kvrpcpb.ScanRequest:
+		r := req.Req.(*kvrpcpb.ScanRequest)
+		if r != nil {
+			firstKey = r.StartKey
+		}
+	case *kvrpcpb.PrewriteRequest:
+		r := req.Req.(*kvrpcpb.PrewriteRequest)
+		if r != nil && len(r.Mutations) > 0 {
+			if mutation := r.Mutations[0]; mutation != nil {
+				firstKey = mutation.Key
+			}
+		}
+	case *kvrpcpb.CommitRequest:
+		r := req.Req.(*kvrpcpb.CommitRequest)
+		if r != nil && len(r.Keys) > 0 {
+			firstKey = r.Keys[0]
+		}
+	case *kvrpcpb.BatchRollbackRequest:
+		r := req.Req.(*kvrpcpb.BatchRollbackRequest)
+		if r != nil && len(r.Keys) > 0 {
+			firstKey = r.Keys[0]
+		}
+	case *coprocessor.Request:
+		r := req.Req.(*coprocessor.Request)
+		if r != nil && len(r.Ranges) > 0 {
+			if keyRange := r.Ranges[0]; keyRange != nil {
+				firstKey = keyRange.Start
+			}
+		}
+	case *coprocessor.BatchRequest:
+		r := req.Req.(*coprocessor.BatchRequest)
+		if r != nil && len(r.Regions) > 0 {
+			if region := r.Regions[0]; region != nil {
+				if len(region.Ranges) > 0 {
+					if keyRange := region.Ranges[0]; keyRange != nil {
+						firstKey = keyRange.Start
+					}
+				}
+			}
+		}
+	}
+	return
+}
+
+// variables needed by decodeKeyHead.
+//
+// **These variables are copied from tablecodec package to avoid circular dependency.**
+var (
+	tablePrefix     = []byte{'t'}
+	recordPrefixSep = []byte("_r")
+	indexPrefixSep  = []byte("_i")
+	errInvalidKey   = dbterror.ClassXEval.NewStd(errno.ErrInvalidKey)
+)
+
+// decodeKeyHead decodes the key's head and gets the tableID, indexID. isRecordKey is true when is a record key.
+//
+// **This function is copied from tablecodec.DecodeKeyHead to avoid circular dependency.**
+func decodeKeyHead(key []byte) (tableID int64, indexID int64, isRecordKey bool, err error) {
+	isRecordKey = false
```

Comment:
Actually, for the resource group tag, what you really need (minimized) is whether a key is an index key or row key. This seems to be the actual API you need (which is also the most efficient one):

```go
type kvKeyKind int

const unknown kvKeyKind = iota
const rowKey kvKeyKind
const indexKey kvKeyKind

func getKeyKind(key []byte) kvKeyKind {
  // check t prefix
  // check _r and length  -> row key
  // check _i and length  -> index key
  // -> unknown
}
```

BTW even for the logic above, it would be better if they can be put inside tablecodec, as they are really "table codec" things.

Reply:
I implemented a minimized [function](https://github.com/pingcap/tidb/pull/29044/files#diff-0601d5cd0c3ca23cf90091c54956f550e9565709161fd76fb009a8356718c88fR68) for distinguishing row/index. I did not abstract this small piece of code and place it in the `tablecodec`, because we have the following dependencies:

`tablecodec -> stmtctx -> resourcegrouptag `

If we let `resourcegrouptag -> tablecodec`, a circular dependency will occur. In fact, in my opinion, `tablecodec` should be a more basic component, but it depends on the complex `stmtctx`.

Reply:
Got it. This seems to be a problem of tablecodec. Can we improve that part? In my opinion, it would be best if our check key function is implemented in tablecodec, since it can reduce the risk when other people changing the key format -- they will find that resourcegrouptag is depending on it, and they should take care. Otherwise we will be the only person to ensure that resourcegroup is working all the time.

Reply:
`tablecodec` only uses `stmtctx.StatementContext.Timezone`, which seems to be easy to fix. Unfortunately, other components that `tablecodec` depends on also depend on `stmtctx`, such as `types`, `util/codec`, `util/rawcodec`, etc. Fixing all this seems to require a lot of work.

So I created a [subpackage](https://github.com/pingcap/tidb/pull/29044/files#diff-ce7f04651e61f17645eab0998751e45c31a5aa433088d732944efdb575cbec67) named `rowindexcodec` for `tablecodec` (maybe this is not a good name, but I don’t have any ideas yet), so that `rowindexcodec` is separated from `resourcegrouptag` and belongs to `tablecodec`, and there is no circular dependency here.